repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+          case Some(clusterWrapper) =>
+            val routableNode = clusterWrapper.getNextNode
+            val port = if (HttpsProtocol.equalsIgnoreCase(cluster.getProtocol)) routableNode.getHttpsPort else routableNode.getHttpPort
+            Some(new URL(cluster.getProtocol, routableNode.getHostname, port, cluster.getRootPath + uri))
+          case None =>
+            logger.warn("No routable node for domain: " + cluster.getId)
+            None
+        }
+      case _ =>
+        logger.warn(s"Unknown destination type: ${destination.getClass.getName}")
+        None
+    }
+  }
+
+  def buildEndpoint(localhost: Node, localPorts: List[Port], endpoint: DestinationEndpoint, uri: String, request: HttpServletRequest): URL = {
+    val protocol = endpoint.getProtocol
+    val endpointPort = endpoint.getPort
+    val endpointHost = endpoint.getHostname
+    val path = StringUriUtilities.concatUris(endpoint.getRootPath, uri)
+
+    val portUrl = if (!StringUtils.isBlank(protocol)) {
+      val portNum = if (endpointPort <= 0) {
+        localPorts.find(_.getProtocol.equalsIgnoreCase(protocol)) match {
+          case Some(port) => port.getNumber
+          case _ => 0
+        }
+      } else {
+        endpointPort
+      }
+      new Port(protocol, portNum)
+    } else {
+      val port = new Port(request.getScheme, request.getLocalPort)
+      if (localPorts.contains(port)) {
+        port
+      } else {
+        throw new MalformedURLException("Cannot determine destination port.")
+      }
+    }
+    // IF the endpoint hostname is blank, THEN it is local
+    val hostnameUrl = if (StringUtils.isBlank(endpointHost)) localhost.getHostname else endpointHost
+
+    new URL(portUrl.getProtocol, hostnameUrl, portUrl.getNumber, path)
+  }
+
+  def doDispatch(servletRequest: HttpServletRequestWrapper,
+                 servletResponse: HttpServletResponse,
+                 locationUrl: URL,
+                 uri: String,
+                 targetContext: ServletContext,
+                 routeDest: RouteDestination,
+                 dispatcher: RequestDispatcher,
+                 destination: Destination): Unit = {
+    servletRequest.setScheme(locationUrl.getProtocol)
+    servletRequest.setServerName(locationUrl.getHost)
+    servletRequest.setServerPort(locationUrl.getPort)
+    servletRequest.setRequestURI(locationUrl.getPath)
+
+    requestHeaderService.setVia(servletRequest)
+    requestHeaderService.setXForwardedFor(servletRequest)
+    logger.debug("Attempting to route to: {}", locationUrl.getPath)
+    logger.debug("  Using dispatcher for: {}", uri)
+    logger.debug("           Request URL: {}", servletRequest.getRequestURL)
+    logger.debug("           Request URI: {}", servletRequest.getRequestURI)
+    logger.debug("          Context path: {}", targetContext.getContextPath)
+    val startTime = System.currentTimeMillis
+    Try {
+      reportingService.incrementRequestCount(routeDest.getDestinationId)
```

Comment:
What is the reporting service?

Reply:
It's a (very old) lookup and tracking service. It basically provides a way to get a `DestinationInfo` object from a `destinationId` String and a way to interact with that `DestinationInfo` object.

The `DestinationInfo` object itself keeps track of number of responses, response codes, and processing time for a specific destination. So basically, rudimentary metrics. It also provides some convenience methods to add more meaning to the data it tracks, like `getThroughput`.

Oh, and of course, there's plenty of abstraction for the sake of abstraction.

Reply:
What else is using it?
Should these be converted to use the JMX metrics?

Reply:
Probably. That's why we use a more robust metrics library after all. That would consolidate our metrics, and allow them to be reported to more that _just_ JMX.

Oh, and I think `ReposeReport` is the only thing using it.

Reply:
Sounds like we have another story.

Reply:
Story created:
* https://repose.atlassian.net/browse/REP-7180